File: backend/app/schemas/public.py
import re
from datetime import date as date_
from typing import Literal

from pydantic import BaseModel, EmailStr, Field, field_validator

from app.models.enums import BookingStatus

PHONE_RE = re.compile(r"^\+?[0-9 \-()]{7,20}$")


class GuestIn(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    email: EmailStr
    phone: str = Field(min_length=7, max_length=20)

    @field_validator("name")
    @classmethod
    def strip_name(cls, v: str) -> str:
        v = v.strip()
        if not v:
            raise ValueError("Name cannot be blank")
        return v

    @field_validator("phone")
    @classmethod
    def validate_phone(cls, v: str) -> str:
        v = v.strip()
        if not PHONE_RE.match(v):
            raise ValueError("Phone number must be 7-20 digits, optionally with +, spaces, hyphens, or parentheses")
        return v


class BookingCreate(BaseModel):
    property_id: int = Field(gt=0)
    room_type_id: int = Field(gt=0)
    rate_plan_id: int = Field(gt=0)
    check_in: date_
    check_out: date_
    guests_count: int = Field(gt=0, le=20)
    # "manual" is admin-only (see admin/bookings.py) — never accepted from the public API.
    payment_method: Literal["online", "whatsapp"]
    guest: GuestIn


class ConfirmPaymentIn(BaseModel):
    email: EmailStr


class BookingOut(BaseModel):
    id: int
    status: BookingStatus
    source: str
    check_in: date_
    check_out: date_
    guests_count: int
    total_amount: float
    property_id: int
    room_type_id: int


class RatePlanOut(BaseModel):
    id: int
    name: str
    fixed_price: float
    refundable: bool
    includes_breakfast: bool


class AvailabilityDayOut(BaseModel):
    date: date_
    rooms_available: int


class RoomAvailabilityOut(BaseModel):
    room_type_id: int
    days: list[AvailabilityDayOut]
    rate_plans: list[RatePlanOut]

File: backend/src/schemas/public.ts
import { z } from "zod";

export const PHONE_RE = /^\+?[0-9 \-()]{7,20}$/;

export const GuestInSchema = z.object({
  name: z
    .string()
    .min(1)
    .max(120)
    .transform((v) => v.trim())
    .refine((v) => v.length > 0, "Name cannot be blank"),
  email: z.string().email(),
  phone: z
    .string()
    .min(7)
    .max(20)
    .transform((v) => v.trim())
    .refine((v) => PHONE_RE.test(v), "Phone number must be 7-20 digits, optionally with +, spaces, hyphens, or parentheses"),
  id_document_url: z.string().max(2000).nullable().optional().default(null),
});
export type GuestIn = z.infer<typeof GuestInSchema>;

export const BookingCreateSchema = z.object({
  property_id: z.number().int().positive(),
  room_type_id: z.number().int().positive(),
  rate_plan_id: z.number().int().positive(),
  check_in: z.string().regex(/^\d{4}-\d{2}-\d{2}$/, "Invalid date"),
  check_out: z.string().regex(/^\d{4}-\d{2}-\d{2}$/, "Invalid date"),
  guests_count: z.number().int().positive().max(20),
  payment_method: z.enum(["online", "whatsapp"]),
  guest: GuestInSchema,
});
export type BookingCreate = z.infer<typeof BookingCreateSchema>;

export const ConfirmPaymentInSchema = z.object({
  email: z.string().email(),
});

export const BookingCancelInSchema = z.object({
  email: z.string().email(),
});
